### PR TITLE
Add search box to CE. Closes #3329

### LINKED ISSUE #3329: Add search bar to Code Explorer
As the consequence of the discussion from issue #3323, I'm opening a separate issue to request adding a search bar to the CE, similar to an Access feature shown here:

Search bar image

The primary purpose of the search bar is to quickly filter a list. We have a `@Folder` annotation but when loading a existing & large project, it can be painful to have to add annotations on a per-module level just to get some level of organization not possible with the object explorer. 

This is distinct from the existing search tools such as GoTo Implementation or Find all references because the workflow is primarily to filter the list of code explorer to only a subset of modules (which can be various component types) that happens to contain a common word, making it easy to navigate between modules when applying changes to a common set of functionalities, without having to add the `@Folder` annotation.

Additionally, this can be leveraged to support mass assignment of `@Folder`. 

To help remind people, the fill-in text could be made explicit such as `Search Module Names` rather than just `Search` and provide buttons for `Find Symbol` if they need to look inside the modules rather than filtering the Code Explorer.

### PR DESCRIPTION
**Pull request by Hosch250**
untitled

### CODE REVIEW

File: RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml [hunk 1]
```diff
@@ -431,10 +434,13 @@
             </ToolBar>
         </ToolBarTray>
 
-        <controls:EmptyUIRefresh Grid.Row="1" />
+        <TextBox Grid.Row="1" TextChanged="SearchBox_OnTextChanged" VerticalContentAlignment="Center" Name="SearchBox"></TextBox>
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml]: I don't like handling the `TextChanged` event, but I didn't see a command to bind to. Let me know if you see something.

File: RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml [hunk 2]
```diff
@@ -431,10 +434,13 @@
...
+        <Image Grid.Row="1" Source="{StaticResource SearchImage}" HorizontalAlignment="Right" VerticalAlignment="Center" MaxHeight="16" Margin="0,0,1,0" MouseDown="SearchIcon_OnMouseDown" />
```
Comment (Hosch250) [inline RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml]: I have to handle mouse down so clicking the search sends focus to the textbox; I think this is OK.

File: RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml (hunk 1 above)
Comment (retailcoder) [reply RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml]: Why isn't the tag self-closed?

File: RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs [hunk 3]
```diff
@@ -577,6 +577,23 @@ public Visibility EmptyUIRefreshMessageVisibility
             }
         }
 
+        public void Search(IEnumerable<CodeExplorerItemViewModel> nodes, string searchString)
```
Comment (retailcoder) [inline RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs]: This is more than a `Search` though, the name doesn't convey the side-effects the method has - e.g. I'd expect a `Search` method to return the search results or something. Seems `FilterByName` would be more descriptive. From the user's point of view that can remain a "search box" though.

File: RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml (hunk 1 above)
Comment (Hosch250) [reply RetailCoder.VBE/UI/CodeExplorer/CodeExplorerControl.xaml]: Getting much too used to HTML at work...

File: RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs (hunk 3 above)
Comment (Hosch250) [reply RetailCoder.VBE/Navigation/CodeExplorer/CodeExplorerViewModel.cs]: You're right.

**Comment by retailcoder** [comment]
That's a good start, but I think it needs a few more things for a good UX:

- A button to clear the filter/textbox
- Some kind of toggle button that allows selectively applying the filter to modules or members (but not both) - these could use icons for class module and member method, but they would need "Filter modules" and "Filter members" tooltips, respectively.

Perhaps the "search" icon could be a toggle button in the toolbar, that toggles the visibility of the search box? That would free the space for the module/member buttons at the right end.

Also (not for *this* PR though), at this point we should consider having a *Code Explorer Settings* tab for the user to configure default CE toggles (show/hide signature, default folder scheme [by namespace vs by component type... can't find the issue for that atm], and whether or not to show the searchbox by default) - and if we have that then we'll need a little "settings" button on the toolbar to bring up the CE settings tab.

**Comment by Hosch250** [comment]
I don't like the idea of a toggle button much; I designed it to work mainly like VS's. I will add a "x" to clear it. Not sure how the filters would work.

**Comment by retailcoder** [comment]
@Hosch250 Fine by me. Let's wait until someone requests to be able to search for module names instead of members.

**Pull request merged**